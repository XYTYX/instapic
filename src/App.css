@import '~antd/dist/antd.css';

.App {
  text-align: center;
}

.navBarButton {
  margin-left: 5px;
  margin-right: 5px;
}

.feed {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.rowFeed {
  display: flex;
  flex-direction: row;
  align-items: center;
  flex-wrap: wrap;
  justify-content: center;
}

.rowCard {
  max-width: 300px;
}

.image {
  max-width: 700px;
}

.authForm {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.spaced {
  margin-top: 6px
}

.subtitle {
  margin-top: 10px;
}

.card {
  margin-top: 24px;
  display: flex;
  flex-direction: column;
  align-items: center;
}