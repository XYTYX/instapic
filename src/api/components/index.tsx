import React from "react";

interface InputProps {
  type?: string;
  value: string;
  onChange(value: string): void;
}

export function Input(props: InputProps) {
  function onChange(ev: React.ChangeEvent<HTMLInputElement>) {
    props.onChange(ev.target.value);
  }

  const type = props.type ? props.type : "text";

  return <input type={type} value={props.value} onChange={onChange}></input>;
}

interface FormProps {
  children: React.ReactNode;
  onSubmit(): void;
}

export function Form(props: FormProps) {
  function onSubmit() {
    props.onSubmit();
  }

  return <form onSubmit={onSubmit}>{props.children}</form>;
}
