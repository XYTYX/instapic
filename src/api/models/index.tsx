export interface User {
  email: string;
  username: string;
}

export interface Login {
  authorization: string;
}
