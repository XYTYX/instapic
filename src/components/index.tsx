export { NewPostModal } from "./newPostModal";
export { Explore } from "./explore";
export { Feed } from "./feed";
export { ExploreByUsers } from "./exploreByUsers";
export { SingleUserFeed } from "./singleUserFeed";
export { Login } from "./login";
export { Signup } from "./signup";
