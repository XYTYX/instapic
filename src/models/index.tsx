export interface User {
  email: string;
  username: string;
}

export interface AuthToken {
  authorization: string;
}

export interface Post {
  id: string;
  text: string;
  images: Array<Image>;
}

export interface Image {
  path: string;
}
