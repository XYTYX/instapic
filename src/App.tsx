import { createBrowserHistory } from "history";
import React, { useEffect, useRef, useState } from "react";
import { Redirect, Route, Router, Switch } from "react-router";
import { Link } from "react-router-dom";
import { Input, Form, Button, message } from "antd";
import {
  Api,
  ApiImpl,
  DownstreamError,
  LoginFailedError,
  UserAlreadyExistsError,
} from "./api";
import { AuthToken, Post } from "./models";
import HttpClient from "./api/client";
import { Cookies } from "react-cookie";
import "./App.css";
import "./index.css";
import { NewPostModal, Explore } from "./components";

export default function App() {
  const customHistory = createBrowserHistory();

  return (
    <div className="App">
      <Router history={customHistory}>
        <HistoryAwareApp />
      </Router>
    </div>
  );
}

export enum SortBy {
  MOST_RECENT = "most_recent",
  BY_USERS = "by_users",
}

function HistoryAwareApp() {
  const [posts, setPosts] = useState<Array<Post>>([]);
  const cookies = new Cookies();
  const cookieJWT = cookies.get("instapic_jwt");
  const [authToken, setAuthToken] = useState<string>(
    cookieJWT ? cookieJWT : ""
  );
  let httpClient = useRef(new HttpClient(authToken));
  let api = useRef(new ApiImpl(httpClient.current));

  const [visibleModal, setVisibleModal] = useState<boolean>(false);

  function showModal() {
    setVisibleModal(true);
  }

  function hideModal() {
    setVisibleModal(false);
  }

  async function getPosts(sortBy: SortBy) {
    let posts: Array<Post>;
    posts = await api.current.getPosts(sortBy);
    setPosts(posts);
  }

  async function logout() {
    await api.current.logout();
    cookies.remove("instapic_jwt");
    setAuthToken("");
  }

  useEffect(() => {
    httpClient.current.setAuthToken(authToken);
    api.current.setClient(httpClient.current);
  }, [authToken]);

  return (
    <div className="App">
      <p className="spaced"></p>
      <nav>
        {!authToken && (
          <>
            <Link to="/login">
              <Button>Log In</Button>
            </Link>
            <Link to="/signup">
              <Button>Signup</Button>
            </Link>
          </>
        )}
        {authToken && (
          <>
            <Link to="/explore">
              <Button>Explore</Button>
            </Link>
            <Link to="new_post">
              <Button onClick={showModal}>New Post</Button>
              <NewPostModal
                getPosts={getPosts}
                hideModal={hideModal}
                visible={visibleModal}
                api={api.current}
              />
            </Link>
            <Button onClick={logout}>Log Out</Button>
          </>
        )}
      </nav>
      <Switch>
        <Route path="/login">
          {authToken ? (
            <Redirect to="/" />
          ) : (
            <Login api={api.current} setAuthToken={setAuthToken} />
          )}
        </Route>
        <Route path="/signup">
          {authToken ? (
            <Redirect to="/" />
          ) : (
            <Signup api={api.current} setAuthToken={setAuthToken} />
          )}
        </Route>
        <Route path="/explore">
          {!authToken ? (
            <Redirect to="login" />
          ) : (
            <Explore posts={posts} getPosts={getPosts} api={api.current} />
          )}
        </Route>
        <Route path="/">
          {!authToken ? (
            <Redirect to="login" />
          ) : (
            <Explore posts={posts} getPosts={getPosts} api={api.current} />
          )}
        </Route>
      </Switch>
    </div>
  );
}

interface LoginProps {
  api: Api;
  setAuthToken(authToken: string): void;
}

function Login(props: LoginProps) {
  async function onSubmit({ email, password }: any) {
    let response: AuthToken;
    try {
      response = await props.api.login(email, password);
      const cookies = new Cookies();
      const d1 = new Date();
      const d2 = new Date();
      d2.setMinutes(d1.getMinutes() + 15);
      cookies.set("instapic_jwt", response.authorization, { expires: d2 });
      props.setAuthToken(response.authorization);
    } catch (e) {
      if (e instanceof LoginFailedError) {
        message.error(
          "Login unsuccessful, check your email and password. Did you mean to sign up?"
        );
      } else if (e instanceof DownstreamError) {
        message.error(
          "Our systems seem to be experiencing issues, please try again later"
        );
      } else {
        throw e;
      }
    }
  }

  return (
    <div className="authForm">
      <h2 className="spaced">Log In</h2>
      <Form onFinish={onSubmit}>
        <Form.Item
          required
          label="Email"
          name="email"
          rules={[
            { required: true, message: "Please enter an email" },
            { type: "email" },
          ]}
        >
          <Input />
        </Form.Item>
        <Form.Item
          required
          label="Password"
          name="password"
          rules={[{ required: true, message: "Please enter a password" }]}
        >
          <Input.Password />
        </Form.Item>
        <button>Log In</button>
      </Form>
    </div>
  );
}

interface SignupProps {
  api: Api;
  setAuthToken(authToken: string): void;
}

function Signup(props: SignupProps) {
  async function onSubmit({ email, username, password }: any) {
    let response: AuthToken;
    try {
      response = await props.api.signup(email, username, password);
      props.setAuthToken(response?.authorization);
    } catch (e) {
      if (e instanceof UserAlreadyExistsError) {
        message.error(
          "That email or username already exists, did you mean to log in?"
        );
      } else {
        throw e;
      }
    }
  }

  return (
    <div className="authForm">
      <h2 className="spaced">Sign Up</h2>
      <Form onFinish={onSubmit}>
        <Form.Item
          required
          label="Email"
          name="email"
          rules={[
            { required: true, message: "Please enter an email" },
            { type: "email" },
          ]}
        >
          <Input />
        </Form.Item>
        <Form.Item
          required
          label="Username"
          name="username"
          rules={[{ required: true, message: "Please enter a username" }]}
        >
          <Input />
        </Form.Item>
        <Form.Item
          required
          label="Password"
          name="password"
          rules={[
            { required: true, message: "Please enter a password" },
            {
              min: 8,
              message:
                "Make sure your password is at least 8 characters in length",
            },
          ]}
        >
          <Input.Password />
        </Form.Item>
        <button>Sign Up</button>
      </Form>
    </div>
  );
}
