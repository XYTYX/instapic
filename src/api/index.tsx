import HttpClient, {
  ConflictError,
  InternalServerError,
  UnauthorizedError,
} from "./client";
import { AuthToken, Post } from "../models";
import { SortBy } from "../App";

export interface Api {
  login(email: string, password: string): Promise<AuthToken>;
  signup(email: string, username: string, password: string): Promise<AuthToken>;
  logout(): Promise<void>;
  getPosts(sortBy: SortBy): Promise<Array<Post>>;
  newPost(file: File | Blob, text: string): Promise<Post>;
}

export class ApiImpl implements Api {
  constructor(client: HttpClient) {
    this._client = client;
  }

  setClient(client: HttpClient) {
    this._client = client;
  }

  async logout(): Promise<void> {
    const path = "/v1/auth/logout";
    let result: Response;

    try {
      result = await this._client.doPostJson(path, {});
    } catch (e) {
      throw e;
    }
    return;
  }

  async newPost(file: File | Blob, text: string): Promise<Post> {
    const path = "/v1/post";
    let result: Response;
    const formData = new FormData();
    formData.append("file", file);
    formData.append("text", text);
    try {
      result = await this._client.doPostFormData(path, formData);
    } catch (e) {
      throw e;
    }
    return result.json()!!;
  }

  async login(email: string, password: string): Promise<AuthToken> {
    const url = "/v1/auth/login";
    let result: Response;
    try {
      result = await this._client.doPostJson(url, {
        email: email,
        password: password,
      });
      return result.json()!!;
    } catch (e) {
      switch (e) {
        case e instanceof UnauthorizedError: {
          throw new LoginFailedError();
        }
        case e instanceof InternalServerError: {
          throw new DownstreamError();
        }
        default: {
          throw e;
        }
      }
    }
  }

  async signup(
    email: string,
    username: string,
    password: string
  ): Promise<AuthToken> {
    const url = "/v1/user";
    let result: Response;
    try {
      result = await this._client.doPostJson(url, {
        email: email,
        username: username,
        password: password,
      });
      return result.json()!!;
    } catch (e) {
      if (e instanceof ConflictError) {
        throw new UserAlreadyExistsError();
      } else {
        throw e;
      }
    }
  }

  async getPosts(sortBy: SortBy): Promise<Array<Post>> {
    let result: Response;

    const queryParams = {
      sort_by: sortBy,
    };

    try {
      result = await this._client.doGet("/v1/post", queryParams);
      return result.json();
    } catch (e) {
      throw e;
    }
  }

  private _client: HttpClient;
}

export class LoginFailedError extends Error {}
export class DownstreamError extends Error {}
export class UserAlreadyExistsError extends Error {}
