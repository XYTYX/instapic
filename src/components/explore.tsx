import { DownOutlined } from "@ant-design/icons";
import { Button, Dropdown, Menu, Pagination } from "antd";
import React, { useState, useEffect } from "react";
import { Api } from "../api";
import { Post } from "../models";
import { MenuInfo } from "rc-menu/lib/interface";
import { Feed } from "./feed";
import { ExploreByUsers } from ".";

export enum SortBy {
  MOST_RECENT = "most_recent",
  BY_USERS = "by_users",
}

interface ExploreProps {
  api: Api;
  posts: Array<Post>;
  getPosts(sortBy: SortBy, offset: number | null, limit: number | null): void;
  setInherit(inherit: boolean): void;
}

export function Explore(props: ExploreProps) {
  const [sortMethod, setSortMethod] = useState<SortBy>(SortBy.MOST_RECENT);
  const [offset, setOffset] = useState<number>(0);
  // Setting the limit per screen to 3
  const [limit, setLimit] = useState<number>(3);

  useEffect(() => {
    props.getPosts(sortMethod, null, null);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [sortMethod]);

  function renderPosts(posts: Array<Post> | null): React.ReactNode {
    if (posts === null || posts.length === 0) {
      return;
    }

    function onSortMethodClick({ key }: MenuInfo) {
      if (key === SortBy.MOST_RECENT) {
        props.setInherit(false);
        setSortMethod(SortBy.MOST_RECENT);
      } else {
        props.setInherit(true);
        setSortMethod(SortBy.BY_USERS);
      }
    }

    function onNextPage(pageNumber: number, pageSize: number | undefined) {
      const actualPageSize = pageSize ? pageSize : 3;
      setOffset((pageNumber - 1) * actualPageSize);
      setLimit(actualPageSize);
    }

    const dropdownMenu = (
      <Menu onClick={onSortMethodClick}>
        <Menu.Item key={SortBy.MOST_RECENT}>Most Recent</Menu.Item>
        <Menu.Item key={SortBy.BY_USERS}>By User</Menu.Item>
      </Menu>
    );

    // If the user decides to sort by most recently uploaded, we paginate to improve performance
    return (
      <>
        <Dropdown overlay={dropdownMenu} trigger={["click"]}>
          <Button type="text">
            Sort By <DownOutlined />
          </Button>
        </Dropdown>
        {sortMethod === SortBy.MOST_RECENT ? (
          <>
            <Feed posts={posts.slice(offset, offset!! + limit!!)} />
            <Pagination
              style={{ marginTop: "12px", marginBottom: "12px" }}
              onChange={onNextPage}
              defaultCurrent={1}
              total={posts.length}
              pageSize={limit}
            />
          </>
        ) : (
          <ExploreByUsers api={props.api} posts={posts} />
        )}
      </>
    );
  }

  return (
    <div className="spaced">
      <h1 className="bordered" style={{ marginBottom: 0 }}>
        Explore
      </h1>
      {renderPosts(props.posts)}
    </div>
  );
}
