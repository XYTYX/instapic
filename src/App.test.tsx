import React from "react";
import { act, render } from "@testing-library/react";
import App from "./App";
import { Explore, ExploreByUsers, Feed } from "./components";
import renderer from "react-test-renderer";
import { shallow, configure, mount } from "enzyme";
import { Api } from "./api";
import { SortBy } from "./components/explore";
import { Post, User } from "./models";
import Adapter from "enzyme-adapter-react-16";
import { DownOutlined } from "@ant-design/icons";
import { Button, Image } from "antd";

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

configure({ adapter: new Adapter() });

describe("<App />", () => {
  it("app renders correctly", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("renders login when no authtoken", () => {
    const { getAllByText } = render(<App />);
    const button = getAllByText(/Log In/i)[0];
    expect(button).toContainHTML("<span>Log In</span>");
  });

  test("does not render explore when no authtoken", () => {
    const { queryByText } = render(<App />);
    const explore = queryByText(/Explore/i);
    expect(explore).toBeNull();
  });
});

describe("<Feed />", () => {
  it("feed renders correctly", () => {
    const tree = renderer
      .create(
        <Feed
          posts={[
            {
              id: "123",
              text: "text",
              images: [{ full_src: "" }],
              created_on: "123",
              user_public_id: "123",
            },
          ]}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("renders single card text when given properly", () => {
    const { getByText } = render(
      <Feed
        posts={[
          {
            id: "123",
            text: "text",
            images: [{ full_src: "" }],
            created_on: "123",
            user_public_id: "123",
          },
        ]}
      />
    );
    const linkElement = getByText(/text/i);
    expect(linkElement).toBeVisible();
  });
});

function getPostMock(id: string): Post {
  return {
    id: id,
    text: "text",
    images: [{ full_src: "full_src" }],
    created_on: "2020 11 29",
    user_public_id: "user_public_id",
  };
}

function getUserMock(it: string): User {
  return {
    email: it,
    username: "username",
    public_id: "public_id",
  };
}

function getPostsMock() {
  const inner = (
    sortBy: SortBy,
    offset: number | null,
    limit: number | null
  ) => {};
  return inner;
}

const apiMock: Api = {
  login: jest.fn((email, password) =>
    Promise.resolve({ authorization: "authorization" })
  ),
  signup: jest.fn((email, username, password) =>
    Promise.resolve({ authorization: "authorization" })
  ),
  logout: jest.fn(() => Promise.resolve()),
  getPosts: jest.fn((sortBy, offset, limit) =>
    Promise.resolve([getPostMock("1"), getPostMock("2")])
  ),
  newPost: jest.fn((file, text) => Promise.resolve(getPostMock("newPost"))),
  getAllUsers: jest.fn(() => Promise.resolve([getUserMock("a@c")])),
  getUser: jest.fn((publicId) => Promise.resolve(getUserMock(publicId))),
  getPostsByUser: jest.fn((userPublicId) =>
    Promise.resolve([getPostMock("1"), getPostMock("2")])
  ),
};

describe("<Explore />", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Explore
          api={apiMock}
          posts={[getPostMock("1")]}
          getPosts={getPostsMock}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders Explore header even if no posts", () => {
    const wrapper = shallow(
      <Explore api={apiMock} posts={[]} getPosts={getPostsMock()} />
    );
    expect(
      wrapper.contains(<h2 style={{ marginBottom: 0 }}>Explore</h2>)
    ).toBeTruthy();
  });

  it("renders sortBy selector when posts are not empty", () => {
    const wrapper = shallow(
      <Explore
        api={apiMock}
        posts={[getPostMock("1")]}
        getPosts={getPostsMock()}
      />
    );
    expect(
      wrapper.contains(
        <Button type="text">
          Sort By <DownOutlined />
        </Button>
      )
    ).toBeTruthy();
  });

  it("does not render SortBy selector when posts are empty", () => {
    const wrapper = shallow(
      <Explore api={apiMock} posts={[]} getPosts={getPostsMock()} />
    );
    expect(
      wrapper.contains(
        <Button type="text">
          Sort By <DownOutlined />
        </Button>
      )
    ).toBeFalsy();
  });

  it("renders feed when SortBy is by recency", () => {
    const wrapper = shallow(
      <Explore
        api={apiMock}
        posts={[getPostMock("1")]}
        getPosts={getPostsMock()}
      />
    );

    expect(wrapper.contains(<Feed posts={[getPostMock("1")]} />)).toBeTruthy();
  });

  test.skip("renders ExploreByUsers when SortBy is by user", () => {
    const wrapper = shallow(
      <Explore
        api={apiMock}
        posts={[getPostMock("1")]}
        getPosts={getPostsMock()}
      />
    );

    console.log(wrapper.find("Button").simulate("click").debug());
    expect(
      wrapper.contains(
        <ExploreByUsers api={apiMock} posts={[getPostMock("1")]} />
      )
    ).toBeTruthy();
  });
});

describe("<ExploreByUsers />", () => {
  it("renders images when posts are not empty", async () => {
    const wrapper = mount(
      <ExploreByUsers api={apiMock} posts={[getPostMock("1")]} />
    );

    await act(
      () =>
        new Promise<void>((resolve) => {
          setImmediate(() => {
            wrapper.update();
            resolve();
          });
        })
    );

    expect(
      wrapper.contains(
        <Image
          className="image"
          src="http://i2.wp.com/juju.com.hk/wp-content/uploads/2019/11/placeholder.png"
          alt="image"
        />
      )
    ).toBeTruthy();
  });

  it("does not render images when there are no users", async () => {
    var { getAllUsers, ...rest } = apiMock;
    const noUserApiMock = {
      ...rest,
      getAllUsers: jest.fn(() => Promise.resolve([])),
    };
    const wrapper = mount(<ExploreByUsers api={noUserApiMock} posts={[]} />);

    await act(
      () =>
        new Promise<void>((resolve) => {
          setImmediate(() => {
            wrapper.update();
            resolve();
          });
        })
    );

    expect(
      wrapper.contains(
        <Image
          className="image"
          src="http://i2.wp.com/juju.com.hk/wp-content/uploads/2019/11/placeholder.png"
          alt="image"
        />
      )
    ).toBeFalsy();
  });
});
