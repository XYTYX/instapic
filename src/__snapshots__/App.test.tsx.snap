// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> app renders correctly 1`] = `
<div
  className="App"
>
  <div
    className="App gradient"
  >
    <p
      className="spaced"
    />
    <nav>
      <a
        className="navBarButton"
        href="/login"
        onClick={[Function]}
      >
        <button
          className="ant-btn"
          onClick={[Function]}
          type="button"
        >
          <span>
            Log In
          </span>
        </button>
      </a>
      <a
        className="navBarButton"
        href="/signup"
        onClick={[Function]}
      >
        <button
          className="ant-btn"
          onClick={[Function]}
          type="button"
        >
          <span>
            Signup
          </span>
        </button>
      </a>
      
    </nav>
    <div
      className="authForm"
    >
      <h2
        className="spaced"
      >
        Log In
      </h2>
      <form
        className="ant-form ant-form-horizontal"
        onSubmit={[Function]}
      >
        <div
          className="ant-row ant-form-item"
          style={Object {}}
        >
          <div
            className="ant-col ant-form-item-label"
            style={Object {}}
          >
            <label
              className="ant-form-item-required"
              htmlFor="email"
              title="Email"
            >
              Email
            </label>
          </div>
          <div
            className="ant-col ant-form-item-control"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <input
                  className="ant-input"
                  id="email"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          className="ant-row ant-form-item"
          style={Object {}}
        >
          <div
            className="ant-col ant-form-item-label"
            style={Object {}}
          >
            <label
              className="ant-form-item-required"
              htmlFor="password"
              title="Password"
            >
              Password
            </label>
          </div>
          <div
            className="ant-col ant-form-item-control"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <span
                  className="ant-input-affix-wrapper ant-input-password"
                  onMouseUp={[Function]}
                >
                  <input
                    action="click"
                    className="ant-input"
                    id="password"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    style={null}
                    type="password"
                    value=""
                  />
                  <span
                    className="ant-input-suffix"
                  >
                    <span
                      aria-label="eye-invisible"
                      className="anticon anticon-eye-invisible ant-input-password-icon"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseUp={[Function]}
                      role="img"
                      tabIndex={-1}
                    >
                      <svg
                        aria-hidden="true"
                        className=""
                        data-icon="eye-invisible"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M942.2 486.2Q889.47 375.11 816.7 305l-50.88 50.88C807.31 395.53 843.45 447.4 874.7 512 791.5 684.2 673.4 766 512 766q-72.67 0-133.87-22.38L323 798.75Q408 838 512 838q288.3 0 430.2-300.3a60.29 60.29 0 000-51.5zm-63.57-320.64L836 122.88a8 8 0 00-11.32 0L715.31 232.2Q624.86 186 512 186q-288.3 0-430.2 300.3a60.3 60.3 0 000 51.5q56.69 119.4 136.5 191.41L112.48 835a8 8 0 000 11.31L155.17 889a8 8 0 0011.31 0l712.15-712.12a8 8 0 000-11.32zM149.3 512C232.6 339.8 350.7 258 512 258c54.54 0 104.13 9.36 149.12 28.39l-70.3 70.3a176 176 0 00-238.13 238.13l-83.42 83.42C223.1 637.49 183.3 582.28 149.3 512zm246.7 0a112.11 112.11 0 01146.2-106.69L401.31 546.2A112 112 0 01396 512z"
                        />
                        <path
                          d="M508 624c-3.46 0-6.87-.16-10.25-.47l-52.82 52.82a176.09 176.09 0 00227.42-227.42l-52.82 52.82c.31 3.38.47 6.79.47 10.25a111.94 111.94 0 01-112 112z"
                        />
                      </svg>
                    </span>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
        <button>
          Log In
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`<Explore /> renders correctly 1`] = `
<div
  className="spaced"
>
  <h2
    style={
      Object {
        "marginBottom": 0,
      }
    }
  >
    Explore
  </h2>
  <button
    className="ant-btn ant-btn-text ant-dropdown-trigger"
    onClick={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    type="button"
  >
    <span>
      Sort By 
    </span>
    <span
      aria-label="down"
      className="anticon anticon-down"
      role="img"
    >
      <svg
        aria-hidden="true"
        className=""
        data-icon="down"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
        />
      </svg>
    </span>
  </button>
  <div
    className="feed"
  >
    <div
      className="ant-card ant-card-bordered card"
    >
      <div
        className="ant-card-body"
        style={Object {}}
      >
        <div
          className="ant-image image"
          onClick={[Function]}
          style={
            Object {
              "height": undefined,
              "width": undefined,
            }
          }
        >
          <img
            alt="image"
            className="ant-image-img"
            onError={[Function]}
            onLoad={[Function]}
            src="full_src"
          />
        </div>
        <div
          className="ant-card-meta"
          style={
            Object {
              "marginTop": "16px",
            }
          }
        >
          <div
            className="ant-card-meta-detail"
          >
            <div
              className="ant-card-meta-title"
            >
              text
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <ul
    className="ant-pagination"
    style={
      Object {
        "marginBottom": "12px",
        "marginTop": "12px",
      }
    }
    unselectable="unselectable"
  >
    <li
      aria-disabled={true}
      className="ant-pagination-prev ant-pagination-disabled"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex={null}
      title="Previous Page"
    >
      <button
        className="ant-pagination-item-link"
        disabled={true}
        tabIndex={-1}
        type="button"
      >
        <span
          aria-label="left"
          className="anticon anticon-left"
          role="img"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="left"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
            />
          </svg>
        </span>
      </button>
    </li>
    <li
      className="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex="0"
      title={1}
    >
      <a
        rel="nofollow"
      >
        1
      </a>
    </li>
    <li
      aria-disabled={true}
      className="ant-pagination-next ant-pagination-disabled"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex={null}
      title="Next Page"
    >
      <button
        className="ant-pagination-item-link"
        disabled={true}
        tabIndex={-1}
        type="button"
      >
        <span
          aria-label="right"
          className="anticon anticon-right"
          role="img"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="right"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
            />
          </svg>
        </span>
      </button>
    </li>
  </ul>
</div>
`;

exports[`<Feed /> feed renders correctly 1`] = `
<div
  className="feed"
>
  <div
    className="ant-card ant-card-bordered card"
  >
    <div
      className="ant-card-body"
      style={Object {}}
    >
      <div
        className="ant-image image"
        onClick={[Function]}
        style={
          Object {
            "height": undefined,
            "width": undefined,
          }
        }
      >
        <img
          alt="image"
          className="ant-image-img"
          onError={[Function]}
          onLoad={[Function]}
          src=""
        />
      </div>
      <div
        className="ant-card-meta"
        style={
          Object {
            "marginTop": "16px",
          }
        }
      >
        <div
          className="ant-card-meta-detail"
        >
          <div
            className="ant-card-meta-title"
          >
            text
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
